### PR TITLE
Added add-c to include additional compilation units in the compiler i…

### PR DESCRIPTION
**Pull request by jacereda**
…nvocation.

### CODE REVIEW

**Comment by jacereda** [comment]
The name is somewhat misleading, since you can use it to include C++ or ObjC modules.

**Comment by jacereda** [comment]
Maybe the additional compilation units should just be a `Project` field?

**Comment by eriksvedang** [comment]
> Maybe the additional compilation units should just be a Project field?

Yeah I think that makes sense!

**Comment by TimDeve** [comment]
Oh, I've been wanting this for a while, your way makes a lot more sense than what I had in mind ^^

**Comment by hellerve** [comment]
> How about `add-native-source` for a name?

I feel like `native` is a bit of an overloaded word, and keeping `c` somewhere in the name might make sense?

**Comment by TimDeve** [comment]
I gave this a try and it seems to be outputting path relative to the source file, would there be a way for this to output absolute path instead so it can be used in libraries?

**Comment by jacereda** [comment]
@TimDeve can you check if this works better for you?

**Comment by TimDeve** [comment]
Thanks for taking the time to update it. The latest change are closer but still wouldn't work for libraries. Here's a quick program I made to demonstrate:
```clojure
(defndynamic test-add-c [relpath]
  (eval (Dynamic.String.concat [(Dynamic.String.absolute (Dynamic.String.directory (file)))
                                "/"
                                relpath])))

(defn main [] (println* (test-add-c "something.c")))
```

It works if where you're invoking the compiler is where the file is:
```sh
~/d/s/file ❯❯❯ ls
main.carp   out         something.c
~/d/s/file ❯❯❯ carp -x main.carp
/Users/tim/dev/sandbox/file/something.c
```

But not if the source file is in a different folder than where the compiler is invoked (in the case of library for example):
```sh
~/d/s/file ❯❯❯ cd ..
~/d/sandbox ❯❯❯ ls file
main.carp   out         something.c
~/d/sandbox ❯❯❯ carp -x file/main.carp
/Users/tim/dev/sandbox/something.c
```

Ideally it would work like `relative-include` which creates an absolute path from the path of the carp source file.

It would be neat to have a `Dynamic.Path.current-file` utility, as it could be reused for tons of different things.

**Comment by jacereda** [comment]
@TimDeve I guess you're right, I thought `file` was `current-file`.

As for `Dynamic.Path` I guess we need that. In particular, the `absolute` in this PR probably belongs there.

**Comment by jacereda** [comment]
@TimDeve yet another attempt. This time I implemented a `load-stack` command because I didn't find a better way. It also adds the `Dynamic.Path` module.

**Comment by jacereda** [comment]
And looks like I messed it with another change somehow. 

I had a stash for cloning modules with submodules. Do you want me to submit that as another PR?

**Comment by TimDeve** [comment]
Thanks for taking the time to tweak this. It's awesome, it'll allow me to put stuff in libraries when I couldn't before.

File: src/Eval.hs [hunk 1]
```diff
@@ -702,9 +707,7 @@ loadInternal ctx xobj path i reloadMode = do
         _ <- liftIO $ setCurrentDirectory cur
         doGitLoad path fpath
       else do
-        _ <- liftIO $ readProcessWithExitCode "git" ["init"] ""
-        _ <- liftIO $ readProcessWithExitCode "git" ["remote", "add", "origin", path] ""
-        (x0, _, stderr0) <- liftIO $ readProcessWithExitCode "git" ["fetch", "--all", "--tags"] ""
+        (x0, _, stderr0) <- liftIO $ readProcessWithExitCode "git" ["clone", "--recurse-submodules", path, "."] ""
```
Comment (eriksvedang) [inline src/Eval.hs:710]: Is this change equivalent?

File: app/Main.hs [hunk 2]
```diff
@@ -70,6 +70,8 @@ defaultProject =
           , projectGenerateOnly = False
           , projectForceReload = False
           , projectPkgConfigFlags = []
+          , projectCModules = []
+          , projectLoadStack = []
```
Comment (eriksvedang) [inline app/Main.hs:74]: What's the purpose of this?

File: src/Eval.hs (hunk 1 above)
Comment (hellerve) [reply src/Eval.hs:710]: This actually assumes that this is cloned for the first time. It shouldn’t be equivalent?

Comment (eriksvedang) [reply src/Eval.hs:710]: That's what I thought too...

Comment (jacereda) [reply src/Eval.hs:710]: That was accidentally included in this PR, I'll send a separate PR to checkout submodules.

File: app/Main.hs (hunk 2 above)
Comment (jacereda) [reply app/Main.hs:74]: `projectCModules` is the list of modules to be included in the compilation command.
`projectLoadStack` is the list of `.carp` files being loaded so far. It's used in `Dynamic.carp:current-file` to get the file currently being loaded.

**Comment by jacereda** [comment]
Any idea regarding the MacOS build failure?

**Comment by hellerve** [comment]
This has been happening since (at least #903). Since we didn‘t change anything in the CI pipeline I assume this is related to Github CI and Haskell actions somehow, but I haven‘t had time to dig into it yet.

**Pull request merged**